### PR TITLE
Underground optimization project

### PR DESCRIPTION
**Pull request by cableman**
#### Link to ticket

No ticket... rough mode enabled.

#### Description

* Coverted index event into message handler
* Remove FOS ES bundle.
* Move indexing to redis/ES/xx into message system
* Create interfaces to enable swaping out index backend.

#### Screenshot of the result

Nothing to see here...

#### Checklist

- [x] My code is covered by test cases.
- [x] My code passes our test (all our tests).
- [x] My code passes our static analysis suite.
- [x] My code passes our continuous integration process.

If your code does not pass all the requirements on the checklist you have to add a comment explaining why this change 
should be exempt from the list.

#### Additional comments or questions

Linked to https://github.com/danskernesdigitalebibliotek/ddb-cover-service/pull/136

### CODE REVIEW

File: src/Service/Indexing/IndexingElasticService.php [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+    /**
+     * Get client to communicate with ES.
+     *
+     * @return client
+     *   ElasticSearch client
+     *
+     * @throws SearchIndexException
+     */
+    private function getClient(): Client
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:141]: You can configure the `Client` as a service and just inject it:
```
parameters:
    es_config: {'hosts': ['http://es:9200']}

services:
    # lots of service definitions in between    

    Elasticsearch\ClientBuilder: ~

    Elasticsearch\Client:
        factory:   ['@Elasticsearch\ClientBuilder', fromConfig]
        arguments: ['%es_config%']
```
(See: 
- https://stefan-poeltl.medium.com/symfony-meets-elasticsearch-implement-a-search-as-you-type-feature-307e2244f078
- https://github.com/elastic/elasticsearch-php/blob/e55c738d2592d5a121a93f6c3fdc41af5984925b/src/ClientBuilder.php#L156 )

File: src/Service/Indexing/IndexingElasticService.php [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+     *   Name of the index to refresh
+     *
+     * @throws SearchIndexException
+     */
+    private function refreshIndex(string $indexName): void
+    {
+        try {
+            $this->getClient()->indices()->refresh(['index' => $indexName]);
+        } catch (SearchIndexException|ClientResponseException|ServerResponseException $e) {
+            throw new SearchIndexException('Unable to create new index', (int) $e->getCode());
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:169]: As above. Consider adding original exception as `$previous`

File: src/Service/Indexing/IndexingElasticService.php [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+     * @throws SearchIndexException
+     */
+    private function getCurrentActiveIndexName(): string
+    {
+        $client = $this->getClient();
+
+        try {
+            $response = $client->indices()->getAlias(['name' => $this->indexAliasName]);
+        } catch (ClientResponseException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:188]: As above. Consider adding original exception as `$previous`

File: src/Service/Indexing/IndexingElasticService.php [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+                            ],
+                            'height' => [
+                                'type' => 'integer',
+                            ],
+                        ],
+                    ],
+                ],
+            ]);
+        } catch (ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:246]: As above. Consider adding original exception as `$previous`

File: src/Service/Indexing/IndexingElasticService.php [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+     */
+    public function bulkAdd(array $items): void
+    {
+        if (!isset($this->newIndexName)) {
+            $this->newIndexName = $this->indexAliasName.'_'.date('Y-m-d-His');
+            $this->createIndex($this->newIndexName);
+        }
+
+        $params = [];
+        foreach ($items as $item) {
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:84]: Should it be the responsibility of the service or of the caller to split this in sizeable chunks?  
https://www.elastic.co/guide/en/elasticsearch/reference/current/tune-for-indexing-speed.html#_use_bulk_requests

**Review by turegjorup — CHANGES_REQUESTED** [review]
We should consider carefully what we can do to make `index/db/coverstore` operations as atomic as possible. This was abstracted away by `FOSElastica`before but with our own service we need to consider this carefully.

File: baseline.xml [hunk 6]
```diff
@@ -68,16 +68,40 @@
...
-    <PossiblyNullReference occurrences="1">
-      <code>getVendor</code>
-    </PossiblyNullReference>
-  </file>
   <file src="src/Exception/XmlReaderException.php">
     <UndefinedClass occurrences="1">
       <code>Throwable</code>
     </UndefinedClass>
   </file>
+  <file src="src/Message/IndexMessage.php">
```
[the remark this answers is not in the document]
Comment (cableman) [reply baseline.xml]: Tried to fix the once that makes sens now, but not all as they can be fixed in php 8 with null-safe operator.

File: config/services.yaml [hunk 7]
```diff
@@ -82,3 +83,15 @@ services:
     App\Service\VendorService\TheMovieDatabase\TheMovieDatabaseApiService:
         arguments:
             $apiKey: '%env(THEMOVIEDATABASE_APIKEY)%'
+
+    App\Service\PopulateService:
```
[the remark this answers is not in the document]
Comment (cableman) [reply config/services.yaml]: Fixed :-D

File: src/Service/Indexing/IndexItem.php [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+    {
+        $this->height = $height;
+
+        return $this;
+    }
+
+    public function toArray(): array
+    {
+        return [
+            'isIdentifier' => $this->getIsIdentifier(),
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexItem.php:102]: This is not for debugging to convert the object into the data sendt to search as index data

File: src/Service/Indexing/IndexingElasticService.php [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+        try {
+            $response = $this->getClient()->index($params);
+            $this->refreshIndex($this->indexAliasName);
+        } catch (SearchIndexException|ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
+        }
+
+        if (201 !== $response->getStatusCode()) {
+            throw new SearchIndexException('Unable to create new index', $response->getStatusCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:47]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+
+        try {
+            $response = $this->getClient()->delete($params);
+            $this->refreshIndex($this->indexAliasName);
+        } catch (SearchIndexException|ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
+        }
+
+        if (200 !== $response->getStatusCode()) {
+            throw new SearchIndexException('Unable to create new index', $response->getStatusCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:69]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 11]
```diff
@@ -0,0 +1,253 @@
...
+            'index' => $this->indexAliasName,
+            'id' => $item->getId(),
+            'body' => $item->toArray(),
+        ];
+
+        try {
+            $response = $this->getClient()->index($params);
+            $this->refreshIndex($this->indexAliasName);
+        } catch (SearchIndexException|ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:43]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 12]
```diff
@@ -0,0 +1,253 @@
...
+                ],
+            ];
+
+            $params['body'][] = $item->toArray();
+        }
+
+        try {
+            $this->getClient()->bulk($params);
+        } catch (SearchIndexException|ClientResponseException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:99]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        $params = [
+            'index' => $this->indexAliasName,
+            'id' => $id,
+        ];
+
+        try {
+            $response = $this->getClient()->delete($params);
+            $this->refreshIndex($this->indexAliasName);
+        } catch (SearchIndexException|ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:65]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 14]
```diff
@@ -0,0 +1,253 @@
...
+                                'index' => $this->newIndexName,
+                                'alias' => $this->indexAliasName,
+                            ],
+                        ],
+                    ],
+                ],
+            ]);
+            $client->indices()->delete(['index' => $existingIndexName]);
+        } catch (ClientResponseException|MissingParameterException|ServerResponseException $e) {
+            throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:129]: Fixed

File: src/Service/Indexing/IndexingElasticService.php [hunk 15]
```diff
@@ -0,0 +1,253 @@
...
+     *
+     * @throws SearchIndexException
+     */
+    private function getClient(): Client
+    {
+        if (!isset($this->client)) {
+            try {
+                $client = ClientBuilder::create()->setHosts([$this->hostUrl])->build();
+            } catch (AuthenticationException $e) {
+                throw new SearchIndexException($e->getMessage(), (int) $e->getCode());
```
[the remark this answers is not in the document]
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:147]: Fixed

File: src/Service/Indexing/IndexingElasticService.php (hunk 2 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php]: Fixed

File: src/Service/Indexing/IndexingElasticService.php (hunk 3 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php]: Fixed

File: src/Service/Indexing/IndexingElasticService.php (hunk 4 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php]: Fixed

File: src/Service/Indexing/IndexingElasticService.php (hunk 5 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:84]: As we talked about i think the caller as that is also us.

File: src/Service/Indexing/IndexingElasticService.php (hunk 1 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php]: Fixed

File: README.md [hunk 16]
```diff
@@ -70,6 +70,15 @@ The application logs to ElasticSearch to allow debugging and monitoring. A `stat
 ensure that Elastic chooses the right type for the index fields a dynamic index template must be added to Elastic. This
 can be done with the `app:elastic:create-stats-template` command.
 
+### Low disk
+
+If you have less than 10% disk space left elastic search will not index content and switchs into read-only mode. This
+command will disable this behaviour:
```
Comment (turegjorup) [inline README.md:76]: - Is this really what we want? Thinking there is a reason for ES doing this.
- Do we have alerts set up for low disk space?

File: src/Service/Indexing/IndexingElasticService.php [hunk 17]
```diff
@@ -0,0 +1,222 @@
...
+
+    /**
+     * Switch new index with old by updating alias.
+     *
+     * @throws SearchIndexException
+     */
+    public function switchIndex(): void
+    {
+        $existingIndexName = $this->getCurrentActiveIndexName();
+        $this->refreshIndex($this->newIndexName);
```
Comment (turegjorup) [inline src/Service/Indexing/IndexingElasticService.php:108]: `newIndexName` might not be set. The check done in L75 for `bulkAdd` is missing here. This makes a public function depend on call order to function.

We should do a private `getNewIndexName()` function and do null check and set there. Then call that were needed.

Sorry I missed this in the first review

**Review by turegjorup — CHANGES_REQUESTED** [review]
One comment, one required fix (`newIndexName`) - sorry, missed that the first time around

File: src/Service/Indexing/IndexingElasticService.php (hunk 17 above)
Comment (cableman) [reply src/Service/Indexing/IndexingElasticService.php:112]: Fixed

File: README.md (hunk 16 above)
Comment (cableman) [reply README.md:76]: This is only for lokal development... even when I have 80gb free i do this.

**Review by turegjorup — APPROVED** [review]

**Pull request merged**